File: NativeTest/app/src/main/java/com/mcm/nativetest/HSVListener.java
package com.mcm.nativetest;

import android.widget.SeekBar;
import android.widget.TextView;
import org.photonvision.common.util.numbers.IntegerCouple;
import org.photonvision.vision.pipeline.ColoredShapePipelineSettings;

public class HSVListener implements SeekBar.OnSeekBarChangeListener {
    MainActivity mainActivity;
    TextView hueMinText, hueMaxText;
    SeekBar hueMin, hueMax;
    TextView satMinText, satMaxText;
    SeekBar satMin, satMax;
    TextView valueMinText, valueMaxText;
    SeekBar valueMin, valueMax;

    TextView accuracyText;
    SeekBar accuracy;

    public HSVListener(MainActivity mainActivity) {
        this.mainActivity = mainActivity;

        // Hue
        hueMinText = mainActivity.findViewById(R.id.hueMinText);
        hueMin = mainActivity.findViewById(R.id.hueMin);
        hueMaxText = mainActivity.findViewById(R.id.hueMaxText);
        hueMax = mainActivity.findViewById(R.id.hueMax);
        hueMin.setOnSeekBarChangeListener(this);
        hueMax.setOnSeekBarChangeListener(this);

        // Saturation
        satMinText = mainActivity.findViewById(R.id.satMinText);
        satMin = mainActivity.findViewById(R.id.satMin);
        satMaxText = mainActivity.findViewById(R.id.satMaxText);
        satMax = mainActivity.findViewById(R.id.satMax);
        satMin.setOnSeekBarChangeListener(this);
        satMax.setOnSeekBarChangeListener(this);

        // Value
        valueMinText = mainActivity.findViewById(R.id.valueMinText);
        valueMin = mainActivity.findViewById(R.id.valueMin);
        valueMaxText = mainActivity.findViewById(R.id.valueMaxText);
        valueMax = mainActivity.findViewById(R.id.valueMax);
        valueMin.setOnSeekBarChangeListener(this);
        valueMax.setOnSeekBarChangeListener(this);

        // Accuracy percentage

    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        VisionProcessThread thread = mainActivity.getVisionProcess();
        if(thread == null) return;
        if(!fromUser) return;

        ColoredShapePipelineSettings settings = thread.getColoredShapeSettings();
        settings.hsvHue = new IntegerCouple(hueMin.getProgress(), hueMax.getProgress());
        settings.hsvSaturation = new IntegerCouple(satMin.getProgress(), satMax.getProgress());
        settings.hsvValue = new IntegerCouple(valueMin.getProgress(), valueMax.getProgress());

        // hue
        if(settings.hsvHue.getFirst() > settings.hsvHue.getSecond()) {
            int lower = settings.hsvHue.getSecond();
            int upper = settings.hsvHue.getFirst();
            settings.hsvHue = new IntegerCouple(lower, upper);
            hueMin.setProgress(lower);
            hueMax.setProgress(upper);
        }

        // sat
        if(settings.hsvSaturation.getFirst() > settings.hsvSaturation.getSecond()) {
            int lower = settings.hsvSaturation.getSecond();
            int upper = settings.hsvSaturation.getFirst();
            settings.hsvSaturation = new IntegerCouple(lower, upper);
            satMin.setProgress(lower);
            satMax.setProgress(upper);
        }

        // value
        if(settings.hsvValue.getFirst() > settings.hsvValue.getSecond()) {
            int lower = settings.hsvValue.getSecond();
            int upper = settings.hsvValue.getFirst();
            settings.hsvValue = new IntegerCouple(lower, upper);
            valueMin.setProgress(lower);
            valueMax.setProgress(upper);
        }

        hueMinText.setText("Hue Min: " + settings.hsvHue.getFirst());
        hueMaxText.setText("Hue Max: " + settings.hsvHue.getSecond());
        satMinText.setText("Sat Min: " + settings.hsvSaturation.getFirst());
        satMaxText.setText("Sat Max: " + settings.hsvSaturation.getSecond());
        valueMinText.setText("Value Min: " + settings.hsvValue.getFirst());
        valueMaxText.setText("Value Max: " + settings.hsvValue.getSecond());
        thread.setSettings(settings);
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
    }
}

File: NativeTest/app/src/main/java/com/mcm/nativetest/HSVListener.kt
package com.mcm.nativetest

import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import org.photonvision.common.util.numbers.DoubleCouple
import org.photonvision.common.util.numbers.IntegerCouple
import org.photonvision.common.util.numbers.NumberCouple
import org.photonvision.vision.opencv.ContourShape
import org.photonvision.vision.pipeline.ColoredShapePipelineSettings

class HSVListener(private var mainActivity: MainActivity) : OnSeekBarChangeListener {
    private var hueMinText: TextView = mainActivity.findViewById(R.id.hueMinText)
    private var hueMaxText: TextView = mainActivity.findViewById(R.id.hueMaxText)
    private var hueMin: SeekBar = mainActivity.findViewById(R.id.hueMin)
    private var hueMax: SeekBar = mainActivity.findViewById(R.id.hueMax)
    private var satMinText: TextView = mainActivity.findViewById(R.id.satMinText)
    private var satMaxText: TextView = mainActivity.findViewById(R.id.satMaxText)
    private var satMin: SeekBar = mainActivity.findViewById(R.id.satMin)
    private var satMax: SeekBar = mainActivity.findViewById(R.id.satMax)
    private var valueMinText: TextView = mainActivity.findViewById(R.id.valueMinText)
    private var valueMaxText: TextView = mainActivity.findViewById(R.id.valueMaxText)
    private var valueMin: SeekBar = mainActivity.findViewById(R.id.valueMin)
    private var valueMax: SeekBar = mainActivity.findViewById(R.id.valueMax)
    private var accuracyText: TextView = mainActivity.findViewById(R.id.accuracyText)
    private var accuracy: SeekBar = mainActivity.findViewById(R.id.accuracy)

    init {
        // Hue
        hueMin.setOnSeekBarChangeListener(this)
        hueMax.setOnSeekBarChangeListener(this)

        satMin.setOnSeekBarChangeListener(this)
        satMax.setOnSeekBarChangeListener(this)

        valueMin.setOnSeekBarChangeListener(this)
        valueMax.setOnSeekBarChangeListener(this)

        // Accuracy percentage
        accuracy.setOnSeekBarChangeListener(this)
    }

    fun init() {
        // Set initial settings
        val thread = mainActivity.visionProcess
        val settings = thread?.pipelineSettings
        if (settings != null) {
            settings.accuracyPercentage = 75.0
            settings.circleAccuracy = 12
            settings.hsvHue = IntegerCouple(64, 170)
            settings.hsvSaturation = IntegerCouple(199, 255)
            settings.hsvValue = IntegerCouple(179, 255)
            settings.contourArea = DoubleCouple(1.0 / 100.0, 100)
            settings.contourShape = ContourShape.Circle

            setCouple(hueMin, hueMax, settings.hsvHue)
            setCouple(satMin, satMax, settings.hsvSaturation)
            setCouple(valueMin, valueMax, settings.hsvValue)
            accuracy.progress = settings.accuracyPercentage.toInt()

            updateText(settings)
        }
    }

    private fun <T : Number> setCouple(lower: SeekBar, upper: SeekBar, value: NumberCouple<T>) {
        lower.progress = value.first.toInt()
        upper.progress = value.second.toInt()
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        val thread: VisionProcessThread = mainActivity.visionProcess ?: return
        if (!fromUser) return

        val settings = thread.pipelineSettings
        updateSettings(settings)
        updateText(settings)

        thread.setSettings(settings)
    }

    private fun updateSettings(settings: ColoredShapePipelineSettings) {
        settings.hsvHue = IntegerCouple(hueMin.progress, hueMax.progress)
        settings.hsvSaturation = IntegerCouple(satMin.progress, satMax.progress)
        settings.hsvValue = IntegerCouple(valueMin.progress, valueMax.progress)
        settings.accuracyPercentage = accuracy.progress.toDouble()

        // hue
//        if (settings.hsvHue.first > settings.hsvHue.second) {
//            val lower = settings.hsvHue.second
//            val upper = settings.hsvHue.first
//            settings.hsvHue = IntegerCouple(lower, upper)
//            hueMin.progress = lower
//            hueMax.progress = upper
//        }

        // sat
        if (settings.hsvSaturation.first > settings.hsvSaturation.second) {
            val lower = settings.hsvSaturation.second
            val upper = settings.hsvSaturation.first
            settings.hsvSaturation = IntegerCouple(lower, upper)
            satMin.progress = lower
            satMax.progress = upper
        }

        // value
        if (settings.hsvValue.first > settings.hsvValue.second) {
            val lower = settings.hsvValue.second
            val upper = settings.hsvValue.first
            settings.hsvValue = IntegerCouple(lower, upper)
            valueMin.progress = lower
            valueMax.progress = upper
        }
    }

    private fun updateText(settings: ColoredShapePipelineSettings) {
        accuracyText.text = "Accuracy: ${settings.accuracyPercentage}"
        hueMinText.text = "Hue Min: " + settings.hsvHue.first
        hueMaxText.text = "Hue Max: " + settings.hsvHue.second
        satMinText.text = "Sat Min: " + settings.hsvSaturation.first
        satMaxText.text = "Sat Max: " + settings.hsvSaturation.second
        valueMinText.text = "Value Min: " + settings.hsvValue.first
        valueMaxText.text = "Value Max: " + settings.hsvValue.second
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {}
    override fun onStopTrackingTouch(seekBar: SeekBar) {}
}